### PR TITLE
Spanner: Re-create sessions that have been invalidated by the server

### PR DESCRIPTION
**Pull request by olavloite**
Sessions in the session pool that have been invalidated by the server may cause `SessionNotFoundException`s to bubble up to the user, even though the user has no control over the session management. When this occurs, the session should be removed from the SessionPool and a new session should be fetched from the SessionPool without the user having to intervene.
In case of a `SessionNotFoundException` being thrown while using a `TransactionManager`, the client library will replace the session with a new one from the pool, and trigger a restart of the transaction by throwing an `AbortedException` (it is not possible to replace the session without also restarting the transaction).

This change has been initiated on request from the Spanner team.

### CODE REVIEW

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 1]
```diff
@@ -75,68 +76,119 @@ Instant instant() {
...
+                session.markUsed();
+                beforeFirst = false;
+              }
+              if (!ret && isSingleUse) {
+                close();
+              }
+              return ret;
+            } catch (SessionNotFoundException e) {
+              recreateSessionIfPossible(e);
+              replaceDelegate(resultSetSupplier.get());
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: can we just do a `return next()` here instead of adding `while(true)`?

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 2]
```diff
@@ -206,7 +268,16 @@ public Timestamp getReadTimestamp() {
...
     public TransactionContext begin() {
-      return delegate.begin();
+      while (true) {
+        try {
+          TransactionContext res = delegate.begin();
+          session.markUsed();
+          return res;
+        } catch (SessionNotFoundException e) {
+          session.maybeRecreateUnderlyingSession(e);
+          delegate = session.delegate.transactionManager();
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: can you just use `return begin()` instead of adding the `while` loop?

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 3]
```diff
@@ -75,68 +76,119 @@ Instant instant() {
...
+      session.maybeRecreateUnderlyingSession(e);
+      session.markUsed();
+      readContextDelegate = readContextDelegateSupplier.get();
+    }
+
     @Override
     public ResultSet read(
-        String table, KeySet keys, Iterable<String> columns, ReadOption... options) {
-      return wrap(delegate.read(table, keys, columns, options));
+        final String table,
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: Can you please make style changes in a separate PR?

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 4]
```diff
@@ -378,17 +496,33 @@ public ReadOnlyTransaction singleUseReadOnlyTransaction(TimestampBound bound) {
...
         throw e;
       }
     }
 
     @Override
-    public ReadOnlyTransaction readOnlyTransaction(TimestampBound bound) {
+    public ReadOnlyTransaction readOnlyTransaction(final TimestampBound bound) {
       try {
-        return new AutoClosingReadTransaction(delegate.readOnlyTransaction(bound), this, false);
+        return new AutoClosingReadTransaction(
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: This snippet is used 4 times.  Please create a method for this snippet.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java [hunk 5]
```diff
@@ -317,13 +317,17 @@ public DatabaseClient getDatabaseClient(DatabaseId db) {
...
         SessionPool pool = SessionPool.createPool(getOptions(), db, SpannerImpl.this);
-        DatabaseClientImpl dbClient = new DatabaseClientImpl(pool);
+        DatabaseClientImpl dbClient = createDatabaseClient(pool);
         dbClients.put(db, dbClient);
         return dbClient;
       }
     }
   }
 
+  DatabaseClientImpl createDatabaseClient(SessionPool pool) {
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java]: Please make these types of style changes in a different PR.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java (hunk 1 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: I replaced the while(true) loop with an internal method that is can be re-used in the SessionNotFound handler.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java (hunk 2 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: I replaced the while(true) loop with an internal method that is can be re-used in the SessionNotFound handler.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java (hunk 3 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: This change is made automatically by the code formatter, as the method parameters are changed to final, which is needed in order to be able to reference them in the anonymous inner class.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java (hunk 4 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: Added method to reduce code duplication.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java (hunk 5 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java]: I'll keep that in mind for future changes.
For now, I added a missing `@VisibleForTesting` annotation.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 6]
```diff
@@ -75,35 +76,66 @@ Instant instant() {
...
+      } catch (SessionNotFoundException e) {
+        recreateSessionIfPossible(e);
+        res = resultSetSupplier.get();
       }
-      return new ForwardingResultSet(resultSet) {
+      return new ForwardingResultSet(res) {
+        private boolean beforeFirst = true;
+
         @Override
         public boolean next() throws SpannerException {
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: can we change to isNext() and isInternalNext()?

Comment (kolea2) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: nevermind - didn't realize this wasn't a new method

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 7]
```diff
@@ -75,35 +76,66 @@ Instant instant() {
...
 
     private AutoClosingReadContext(
-        ReadContext delegate, PooledSession session, boolean isSingleUse) {
-      this.delegate = delegate;
+        Supplier<T> delegateSupplier, PooledSession session, boolean isSingleUse) {
+      this.readContextDelegateSupplier = delegateSupplier;
       this.session = session;
       this.isSingleUse = isSingleUse;
+      try {
+        this.readContextDelegate = delegateSupplier.get();
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: I would use this.readContextDelegateSupplier

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 8]
```diff
@@ -114,29 +146,56 @@ public boolean next() throws SpannerException {
...
+            AutoClosingReadContext.this.close();
+          }
         }
       };
     }
 
+    private void recreateSessionIfPossible(SessionNotFoundException e) {
+      session.maybeRecreateUnderlyingSession(e);
+      session.markUsed();
+      readContextDelegate = readContextDelegateSupplier.get();
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: can we just call getReadContextDelegate()?

File: google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java [hunk 9]
```diff
@@ -53,6 +58,13 @@ public GceTestEnvConfig() {
     if (!serverUrl.isEmpty()) {
       builder.setHost(serverUrl);
     }
+    if (!credentialsFile.isEmpty()) {
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java]: Can you please clarify what this change is?

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java (hunk 7 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: I'm not sure I understand what you mean. The variable `this.readContextDelegate` is of type `T extends ReadContext`, so I don't think it should be named `this.readContextDelegateSupplier`. Or do you mean something else?

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java (hunk 8 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: No, it's important that `Supplier.get()` is called here, as the `Supplier` should (have the chance to) re-create the `ReadContext`. If the session was re-created at line 157, the `Supplier` must also re-create the `ReadContext` for it to be associated with the new session.

File: google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java (hunk 9 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java]: This variable works in the same way as the serverUrl variable. It is read from a custom environment variable named `spanner.gce.config.credentials_file` (defined in `GceTestEnvConfig#GCE_CREDENTIALS_FILE`) and these credentials are used if this custom environment variable has been set. If it has not been set, the integration test will use the default credentials.
The reason that I have added this option is that when I run the integration tests locally, I use a number of different environments that need different credentials, and changing the default credentials of my local environment is a lot more cumbersome than just passing in a custom system property when running the tests.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java (hunk 7 above)
Comment (kolea2) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: Apologies for not being clear - what I meant was I would have expected this line to be
`this.readContextDelegate = this.readContextDelegateSupplier.get();`
since we just assigned it earlier. Let me know if I'm missing something.

File: google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java (hunk 9 above)
Comment (kolea2) [reply google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java]: Gotcha - since this is already a larger PR, can we move this out into a separate change?

Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java]: I moved the change to a new PR.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 10]
```diff
@@ -289,32 +372,48 @@ private PooledSession(Session delegate) {
...
     }
 
+    @VisibleForTesting
+    void setAllowDelegateRecreation(boolean allow) {
+      allowDelegateRecreation = allow;
+    }
+
+    private void maybeRecreateUnderlyingSession(SessionNotFoundException e) {
+      if (beforeFirstServerCall && allowDelegateRecreation) {
+        delegate = spanner.createSession(db);
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: could we get another session from the pool instead of creating a new one ?

I am afraid that the latency on this request will be too high when we internally retry.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 11]
```diff
@@ -289,32 +372,48 @@ private PooledSession(Session delegate) {
...
+        lastException = null;
+        return internalRunWithSessionRetry(method);
       }
     }
 
-    @Override
-    public long executePartitionedUpdate(Statement stmt) throws SpannerException {
+    private <T> T internalRunWithSessionRetry(Supplier<T> method) {
       try {
+        T res = method.get();
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: should this be after markUsed() ?

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 12]
```diff
@@ -289,32 +372,48 @@ private PooledSession(Session delegate) {
...
-        return delegate.write(mutations);
-      } catch (SpannerException e) {
-        throw lastException = e;
-      } finally {
-        close();
+        return internalRunWithSessionRetry(method);
+      } catch (SessionNotFoundException e) {
+        maybeRecreateUnderlyingSession(e);
+        lastException = null;
+        return internalRunWithSessionRetry(method);
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: there is a chance we got sessionNotFound here as well.. 

have u considered putting this in a loop ? 
Pros: customers never see a SessionNotFound
Cons: high latency, but our timeouts should still take effect.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 13]
```diff
@@ -404,8 +557,15 @@ public TransactionRunner readWriteTransaction() {
         @Nullable
         public <T> T run(TransactionCallable<T> callable) {
           try {
+            T result;
+            try {
+              result = runner.run(callable);
+            } catch (SessionNotFoundException e) {
+              maybeRecreateUnderlyingSession(e);
+              TransactionRunner newRunner = delegate.readWriteTransaction();
+              result = newRunner.run(callable);
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: same as above, this could potentially still see a sessionNotFound (SNF)

**Comment by sduskis** [comment]
@olavloite, it looks like there are conflicts.

**Comment by olavloite** [comment]
@sduskis Yeah, I know. I'll start to rebase it once #4895 has been merged, as that will otherwise also cause further merge conflicts with this change.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java (hunk 1 above)
Comment (kolea2) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: Should the while(true) still be here?

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java [hunk 14]
```diff
@@ -919,6 +1245,24 @@ Session getReadWriteSession() {
...
+  private PooledSession replaceReadSession(SessionNotFoundException e, PooledSession session) {
+    if (!options.isFailIfSessionNotFound() && session.allowReplacing) {
+      closeSessionAsync(session);
+      return getReadSession();
+    } else {
+      throw e;
+    }
+  }
+
+  PooledSession replaceReadWriteSession(SessionNotFoundException e, PooledSession session) {
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: should these have the same visibility?

Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java]: `replaceReadWriteSession` needs at least package visibility, as it is called from `DatabaseClientImpl`. `replaceReadSession` does not need more than private visibility, as it is only called from withing `SessionPool.java`.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java [hunk 15]
```diff
@@ -146,24 +176,48 @@ public TransactionRunner readWriteTransaction() {
...
+  }
+
+  private <T> T runWithSessionRetry(SessionMode sessionMode, Function<Session, T> callable) {
+    PooledSession session =
+        sessionMode == SessionMode.READ ? getReadSession() : getReadWriteSession();
+    while (true) {
+      try {
+        return callable.apply(session);
+      } catch (SessionNotFoundException e) {
+        session = pool.replaceReadWriteSession(e, session);
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: Following up from my last comment, if the session is READ, is this correct, as replaceReadWriteSession eventually returns 'getReadWriteSession()'?

Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: Good point, it wasn't right, even though the method was never called with READ. But that was not correct either, as `executePartitionedUpdate` does not need a read/write session, as it always starts a new transaction (as it needs a special type of transaction).

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java [hunk 16]
```diff
@@ -188,7 +188,7 @@ public long executePartitionedUpdate(final Statement stmt) {
     Span span = tracer.spanBuilder(PARTITION_DML_TRANSACTION).startSpan();
     try (Scope s = tracer.withSpan(span)) {
       return runWithSessionRetry(
-          SessionMode.READ_WRITE,
+          SessionMode.READ, // PartitionedUpdate does not need a prepared tx.
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: Can we do this in a separate PR? Because this PR is quite large, I'd like to separate out the recreating sessions implementation from the refactorings/fixes.

Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: I'll move it to a separate PR later this afternoon.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java [hunk 17]
```diff
@@ -39,11 +42,32 @@
...
+    return pool.getReadSession();
+  }
+
+  @VisibleForTesting
+  PooledSession getReadWriteSession() {
+    return pool.getReadWriteSession();
+  }
+
+  @VisibleForTesting
+  int getNumberOfAvailableWritePreparedSessions() {
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: this is not DBClient's responsibility. I wouldn't expose it here

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java [hunk 18]
```diff
@@ -69,7 +99,7 @@ public Timestamp writeAtLeastOnce(Iterable<Mutation> mutations) throws SpannerEx
   public ReadContext singleUse() {
     Span span = tracer.spanBuilder(READ_ONLY_TRANSACTION).startSpan();
     try (Scope s = tracer.withSpan(span)) {
-      return pool.getReadSession().singleUse();
+      return getReadSession().singleUse();
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: curious whats the behavior with read sessions seeing SNF errors ?

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java [hunk 19]
```diff
@@ -146,24 +176,41 @@ public TransactionRunner readWriteTransaction() {
...
+          });
     } catch (RuntimeException e) {
       TraceUtil.endSpanWithFailure(span, e);
       throw e;
     }
   }
 
+  private <T> T runWithSessionRetry(Function<Session, T> callable) {
+    PooledSession session = getReadWriteSession();
+    while (true) {
```
Comment (snehashah16) [inline google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: Will this respect client side timeout values set ?
If our backend is down, and does not return any sessions -> whats the outcome of the customer application ?

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java (hunk 17 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: Method is now removed and the visibility of the session pool changed to package-private to make it visible for test cases.

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java (hunk 18 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: The SNF is detected at the first actual database request. For read-only transactions (both single use and actual read transactions), that is at the first query. The query is sent to the server at the first call to ResultSet#next(), so the SNF for a read transaction is handled here:
https://github.com/googleapis/google-cloud-java/blob/1b61e287e97cc9cc17333caf50413346edfd00b6/google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java#L129

File: google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java (hunk 19 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java]: The short answer: It will behave the same as now, and the behavior will depend on whether the pool has a session ready to be returned or not. The `runWithSessionRetry` will only retry on the specific condition of a `SessionNotFoundException` being returned.

There are two possible scenarios:
1. The application requests a session and the session pool has a session that it thinks is valid. This session is handed to the application and the application tries to execute the transaction. This will cause the query or begin transaction statement to time out.
2. The pool does not have a session available that can be returned and starts the creation of a new session asynchronously. The requesting thread is placed in a waiter. The create session request will fail and notify the waiter of the error. The waiting thread will return with the create session error.

Both of the above will not be considered a `SessionNotFoundException` by the `runWithSessionRetry` method, and the method will return with an error.
